### PR TITLE
feat: add Maestro recipe

### LINKED ISSUE #56: Add Maestro recipe
Would be great to reuse as much as possible form detox.

### PR DESCRIPTION
**Pull request by km1chno**
Resolves #56

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -82,6 +82,10 @@ The following are **feature flags** that can be used with `--preset` flag (they
     <td style="vertical-align: middle;">--detox</td>
     <td style="vertical-align: middle;">Generate workflow to run Detox e2e tests on every PR</td>
   </tr>
+  <tr>
+    <td style="vertical-align: middle;">--maestro</td>
+    <td style="vertical-align: middle;">Generate workflow to run Maestro e2e tests on every PR</td>
```
Comment (maciekstosio) [inline readme.md:87]: I wonder if we shouldn't extend this readme (or create some extended version somewhere else (np next to each recipe) and write what it does under the hood in more details + known issues like the one with iOS

File: src/types.ts [hunk 2]
```diff
@@ -50,9 +50,14 @@ export interface CycliRecipe {
 
 export type Platform = 'android' | 'ios'
 
+export type Mode = 'debug' | 'release'
```
Comment (maciekstosio) [inline src/types.ts:53]: Shouldn't it be in build recipe file?

File: src/recipes/build.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import { join } from 'path'
+
+const createBuildWorkflowForAndroid = async (
+  toolbox: CycliToolbox,
+  context: ProjectContext,
+  { mode, expo }: { mode: Mode; expo: boolean },
+  workflowProps: Record<string, string> = {}
+): Promise<string> => {
+  const gradleCommands =
+    mode === 'debug' ? 'assembleDebug' : 'assembleRelease assembleAndroidTest'
```
Comment (maciekstosio) [inline src/recipes/build.ts:11]: Maybe `Mode` should be enum

File: src/recipes/build.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    context,
+    workflowProps
+  )
+
+  toolbox.interactive.success(`Created Android ${mode} build workflow.`)
+
+  return workflowFileName
+}
+
+const createBuildWorkflowForIOs = async (
```
Comment (maciekstosio) [inline src/recipes/build.ts:37]: IOS? :D

File: src/templates/build-debug/build-debug-android.ejf [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+}
+%>
+
+name: Build Android Debug App
+
+on:
+  workflow_call:
+    outputs:
+      build-cache-key:
+        description: 'The primary key to used for caching the build'
```
Comment (maciekstosio) [inline src/templates/build-debug/build-debug-android.ejf:21]: without `to` ?

File: src/templates/build-debug/build-debug-ios.ejf [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        run: |
+          mkdir ios-debug-build
+          mv <%= props.pathRelativeToRoot %>/ios/build/Build/Products/Debug-iphonesimulator/<%= props.iOSAppName %>.app ios-debug-build/ios-debug.app
+
+      - name: 📡 Store built app in cache
+        id: ios-debug-save
+        uses: actions/cache/save@v4
+        with:
+          path: ios-debug-build/
+          key: ios-debug-build-${{ needs.lookup-cached-build.outputs.fingerprint }}
```
Comment (maciekstosio) [inline src/templates/build-debug/build-debug-ios.ejf:67]: I wonder if running build for ios won't change the finger print. For example pod install may generate Podfile.lock

Comment (km1chno) [reply src/templates/build-debug/build-debug-ios.ejf:67]: you mean after `expo prebuild`? in this case (the workflow i managed), directories `ios/` and `android` are ignored when calculating fingerprint:

https://github.com/expo/expo/blob/ee198d023415c5d3e0f9ae9676f567badb48821d/packages/expo-updates/utils/src/createFingerprintAsync.ts#L16-L24

File: src/types.ts (hunk 2 above)
Comment (km1chno) [reply src/types.ts]: moved in ef4d52054da13215e5dbe33ba57e407e5fe26939

File: src/recipes/build.ts (hunk 3 above)
Comment (km1chno) [reply src/recipes/build.ts:13]: I'd prefer to stick to types as alternatives of strings to be consistent with types like `PackageManager` and `LockFile`

File: src/recipes/build.ts (hunk 4 above)
Comment (km1chno) [reply src/recipes/build.ts]: changed in ef4d52054da13215e5dbe33ba57e407e5fe26939

File: src/templates/build-debug/build-debug-android.ejf (hunk 5 above)
Comment (km1chno) [reply src/templates/build-debug/build-debug-android.ejf]: thanks, changed in ef4d52054da13215e5dbe33ba57e407e5fe26939

File: src/recipes/build.ts (hunk 3 above)
Comment (maciekstosio) [reply src/recipes/build.ts:13]: You can do something like:

```
const SOMETHING = {
    ONE: 'one',
    TWO: 'two',
} as const

type T = typeof SOMETHING[keyof typeof SOMETHING]
```

Then I think you can use it both ways

File: readme.md (hunk 1 above)
Comment (km1chno) [reply readme.md:87]: This will be done as part of https://github.com/software-mansion-labs/react-native-ci-cli/issues/72

**Review by maciekstosio — DISMISSED** [review]
One minor thing, and we good to go

File: src/templates/maestro/example-flow.ejf [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# example-flow.yaml
+
+appId: <%= props.appId %>
+---
+- launchApp
+- extendedWaitUntil: 
+    visible: "Welcome to React Native"
+    timeout: 120000
```
[the remark this answers is not in the document]
Comment (km1chno) [reply src/templates/maestro/example-flow.ejf]: done in 7c5d7bd962084e118be7d24699a3e06ec54ffc34

**Review by maciekstosio — APPROVED** [review]

**Review by maciekstosio — APPROVED** [review]

**Pull request merged**